metadata/resources_repository_test.go

```diff
@@ -0,0 +1,767 @@
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+	ctx := context.Background()
+
+	// Initialize the test server once for all subtests
+	testServer := newTestMetadataServer(t)
+	defer testServer.Close()
+
+	// Define reusable resources and IDs
+	resources := []ResourceDef{
+		UserDef{
+			Name:       "Featureform",
+			Tags:       Tags{},
+			Properties: Properties{},
+		},
+		UserDef{
+			Name:       "Other",
+			Tags:       Tags{},
+			Properties: Properties{},
+		},
+		ProviderDef{
+			Name:             "mockOffline",
+			Description:      "A mock offline provider",
+			Type:             string(pt.SnowflakeOffline),
+			Software:         "snowflake",
+			Team:             "recommendations",
+			SerializedConfig: []byte(""),
+			Tags:             Tags{},
+			Properties:       Properties{},
+		},
+		SourceDef{
+			Name:        "primarydata",
+			Variant:     "var",
+			Description: "A CSV source but different",
+			Definition: PrimaryDataSource{
+				Location: SQLTable{
+					Name: "mockPrimary",
+				},
+				TimestampColumn: "timestamp",
+			},
+			Owner:      "Featureform",
+			Provider:   "mockOffline",
+			Tags:       Tags{},
+			Properties: Properties{},
+		},
+	}
+
+	resourceIds := make([]ResourceID, 0)
+	for _, res := range resources {
+		resourceIds = append(resourceIds, res.ResourceID())
+	}
+
+	testServer.ResetDatabase()               // Reset DB before test
+	testServer.SeedResources(ctx, resources) // Seed data
+	testServer.SetResourcesReady(ctx, resourceIds)
+
+	t.Run("Delete existing primary", func(t *testing.T) {
+		testServer.ResetDatabase()               // Reset DB before test
+		testServer.SeedResources(ctx, resources) // Seed data
+		testServer.SetResourcesReady(ctx, resourceIds)
+
+		// Attempt to delete the primary
+		err := testServer.repo.MarkForDeletion(ctx, common.ResourceID{
+			Name:    "primarydata",
+			Variant: "var",
+			Type:    common.SOURCE_VARIANT,
+		}, testServer.server.deletionTaskStarter)
+		require.NoError(t, err)
+
+		// Verify the primary is marked for deletion
+		res, err := testServer.repo.Lookup(ctx, ResourceID{
+			Name:    "primarydata",
+			Variant: "var",
+			Type:    SOURCE_VARIANT,
+		}, DeleteLookupOption{DeletedOnly})
+
+		require.NoError(t, err)
+		require.NotNil(t, res)
+		require.Equal(t, res.ID().Name, "primarydata")
+	})
+
+	t.Run("Delete non-existent primary", func(t *testing.T) {
+		testServer.ResetDatabase() // Reset DB before test
+
+		// Attempt to delete a non-existent primary
+		err := testServer.repo.MarkForDeletion(ctx, common.ResourceID{
+			Name: "nonExistentPrimary",
+			Type: common.SOURCE_VARIANT,
+		}, testServer.server.deletionTaskStarter)
+
+		require.Error(t, err) // Expect an error
+	})
+
+	t.Run("Delete primary without READY status", func(t *testing.T) {
+		testServer.ResetDatabase()
+		testServer.SeedResources(ctx, resources)
+
+		// Do NOT set status to READY to simulate invalid state
+
+		// Attempt to delete the primary
```

Comment:
These tests are almost identical to the others above, I'm fine with copy and paste but might be worth taking a second and thinking about how it'd look like if these were all joined